### PR TITLE
Added get headers websocket request.

### PR DESCRIPTION
**Pull request by VolodymyrBabaiev**
In order to improve performance(SC ceased when it more than 50 blocks beyond MC, in order to forge block SC needs to retrieve 50 block from MC, but fails due to timeout) get_block_headers websocket request was implemented. It allows to retrieve up to 50 block headers from MC in one requests.

### CODE REVIEW

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: qa/mc_sc_forging4.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    Start 1 MC nodes and 1 SC node.
+    SC node connected to the first MC node.
+
+Test:
+    - Synchronize MC nodes to the point of SC Creation Block.
+    - Mine 50 MC blocks, then forge SC block, verify MC data inclusion
+    - Mine 150 MC blocks, then forge 2 SC blocks, verify MC data inclusion
+    - Mine 200 MC blocks, then forge SC block, verify MC data inclusion  
+     
+     TODO In tests when SC is more than 50 blocks beyond MC Forger cannot retrieve more than 49 block headers
```
[the remark this answers is not in the document]
Comment (VolodymyrBabaiev) [reply qa/mc_sc_forging4.py:27]: Correct, SC retrieves 50 header hashes one of them is common one.

**Comment by butnaka** [comment]
thank you admin

File: sdk/src/main/scala/com/horizen/forge/MainchainSynchronizer.scala [hunk 2]
```diff
@@ -59,6 +59,22 @@ class MainchainSynchronizer(mainchainNodeChannel: MainchainNodeChannel) {
     }
     references
   }
+
+  def getMainchainBlockHeadersReferences(hashes: Seq[MainchainHeaderHash]): Try[Seq[MainchainHeader]] = Try {
+    val HEADERS_LIMIT = 100 // TODO Analyse amount of headers that can be added to one block.
+    val HEADERS_REQUEST_LIMIT = 25 // TODO Change this value to 50 after changing HEADERS_LIMIT. It has to be 50(as described in doc and implemented in MC)
```
[the remark this answers is not in the document]
Comment (VolodymyrBabaiev) [reply sdk/src/main/scala/com/horizen/forge/MainchainSynchronizer.scala:66]: It's possible to retrieve 50 headers in one request from MC. Limit of headers changed to 25 in order to keep track of correctness of case when getMainchainBlockHeaders makes several header requests to MC. When HEADERS_LIMIT value will be actualized(to maxim possible amount of headers in one block could be included) this value should be changed to 50.

File: sdk/src/main/scala/com/horizen/forge/MainchainSynchronizer.scala [hunk 3]
```diff
@@ -59,6 +59,22 @@ class MainchainSynchronizer(mainchainNodeChannel: MainchainNodeChannel) {
...
+
+    if (hashes.size > HEADERS_LIMIT) throw new IllegalArgumentException("Headers request amount is over the limit(" + HEADERS_LIMIT + ")")
+
+    val strHashes: Seq[String] = hashes.map(hash => BytesUtils.toHexString(hash.data))
+    var headers : Seq[MainchainHeader] = Seq()
+
+    for(group <- strHashes.grouped(HEADERS_REQUEST_LIMIT)) {
+      headers ++= mainchainNodeChannel.getBlockHeaders(group).get
```
[the remark this answers is not in the document]
Comment (VolodymyrBabaiev) [reply sdk/src/main/scala/com/horizen/forge/MainchainSynchronizer.scala:76]: If we put exception handling here in some cases calling function could receive only part of requested headers without being notified of cause of the missing headers. In current implementation calling function receives all requested headers or exception in other case.

File: qa/mc_sc_forging4.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+Configuration:
+    Start 1 MC nodes and 1 SC node.
+    SC node connected to the first MC node.
+
+Test:
+    - Synchronize MC nodes to the point of SC Creation Block.
+    - Mine 50 MC blocks, then forge SC block, verify MC data inclusion
+    - Mine 150 MC blocks, then forge 2 SC blocks, verify MC data inclusion
+    - Mine 200 MC blocks, then forge SC block, verify MC data inclusion
```
Comment (i-Alex) [inline qa/mc_sc_forging4.py:25]: Let's simplify the test flow:
- Mine 200 MC blocks
- Forger N SC blocks and verify MC block headers  and ref data inclusion.
Keeping in mind max number of 49 MC headers per SC block for now.

Later we will update the test case according to the changes in Forger logic.

**Review by i-Alex — CHANGES_REQUESTED** [review]

**Review by i-Alex — APPROVED** [review]

**Review by albertog78 — APPROVED** [review]

**Pull request merged**